utilities/merge_operators/int64subtract_test.cc

```diff
@@ -0,0 +1,257 @@
+
+  s = db_->Get(read_opts_, cf1, "key", &value);
+  ASSERT_OK(s);
+  Slice read_slice(value);
+  int64_t read_value = 0;
+  const bool read_ok = GetVarsignedint64(&read_slice, &read_value);
+
+  db_->DropColumnFamily(cf1);
+  db_->DestroyColumnFamilyHandle(cf1);
+
+  ASSERT_TRUE(read_ok);
+  const int64_t expected = 0 - GetParam();
+  ASSERT_EQ(read_value,
+            expected);  // Merge operators should have been applied on empty db,
+                        // i.e. 0 and then subtracted merge_num.
+}
+
+TEST_P(NonEmptyDbTest, MergeNonEmptyDb) {
+  const int64_t initial_db_num = std::get<0>(GetParam());
+  const int64_t merge_num = std::get<1>(GetParam());
+
+  std::string value;
+
+  ASSERT_OK(OpenDB());
+
+  PutVarsignedint64(&value, initial_db_num);
+  Status s =
+      db_->Put(write_opts_, "key", value);  // Put initial_db_num under key
+  ASSERT_OK(s);
+
+  value.clear();
+
+  PutVarsignedint64(&value, merge_num);
+  s = db_->Merge(write_opts_, "key", value);  // Merging merge_num under key
+  ASSERT_OK(s);
+
+  value.clear();
+
+  s = db_->Get(read_opts_, "key", &value);
+  ASSERT_OK(s);
+  Slice read_slice(value);
+  int64_t read_value = 0;
+  const bool read_ok = GetVarsignedint64(&read_slice, &read_value);
+
+  ASSERT_TRUE(read_ok);
+  const int64_t expected = initial_db_num - merge_num;
+  ASSERT_EQ(read_value,
+            expected);  // Merge operators should have been applied on non-empty
+                        // db, and then merge subtracted merge_num.
+}
+
+TEST_P(NonEmptyDbTest, MergeNonEmptyDbCf) {
+  const int64_t initial_db_num = std::get<0>(GetParam());
+  const int64_t merge_num = std::get<1>(GetParam());
+
+  std::string value;
+
+  ASSERT_OK(OpenDB());
+  ColumnFamilyOptions cf_opts;
+  cf_opts.merge_operator = options_.merge_operator;
+  ColumnFamilyHandle* cf1;
+  Status s = db_->CreateColumnFamily(cf_opts, "cf1", &cf1);
+
+  PutVarsignedint64(&value, initial_db_num);
+  s = db_->Put(write_opts_, cf1, "key", value);  // Put initial_db_num under key
+  ASSERT_OK(s);
+
+  value.clear();
+
+  PutVarsignedint64(&value, merge_num);
+  s = db_->Merge(write_opts_, cf1, "key",
+                 value);  // Merging merge_num under key
+  ASSERT_OK(s);
+
+  value.clear();
+
+  s = db_->Get(read_opts_, cf1, "key", &value);
+  ASSERT_OK(s);
+  Slice read_slice(value);
+  int64_t read_value = 0;
+  const bool read_ok = GetVarsignedint64(&read_slice, &read_value);
+
+  db_->DropColumnFamily(cf1);
+  db_->DestroyColumnFamilyHandle(cf1);
+
+  ASSERT_TRUE(read_ok);
+  const int64_t expected = initial_db_num - merge_num;
+  ASSERT_EQ(read_value,
+            expected);  // Merge operators should have been applied on non-empty
+                        // db, and then merge subtracted merge_num.
+}
+
+TEST_F(Int64SubtractMergeOperatorTest, MergeMultipleValues) {
+  std::string value;
+
+  ASSERT_OK(OpenDB());
+
+  PutVarsignedint64(&value, 123);
+  Status s = db_->Merge(write_opts_, "key", value);  // Merging 123 under key
```

Comment:
I think we should treat the first merge same as a base put. If we don't, we risk getting inconsistent results because compacting the earliest merge to the bottommost level will convert it to a put. Then, this will count as +123 if the compaction to bottommost level has already happened, or -123 if the compaction to bottommost level has not yet happened.

Reply:
> the earliest merge to the bottommost level will convert it to a put

@ajkr This is something I would like to understand and discuss in more detail...

Are you saying that if I do a single `Merge(K1, V1)`, and there is no other preceding op for `K1`. Then at compaction time this will be rewritten to `Put(K1, V1)`.

If that is the case, I am very surprised as it is not mentioned in the documentation: https://github.com/facebook/rocksdb/wiki/Merge-Operator, and there are three places in the API docs that, at least to me, seem to imply the opposite:

1. https://github.com/facebook/rocksdb/blob/master/include/rocksdb/merge_operator.h#L57
2. https://github.com/facebook/rocksdb/blob/master/include/rocksdb/merge_operator.h#L126
3. https://github.com/facebook/rocksdb/blob/master/include/rocksdb/merge_operator.h#L208

In addition that would seem to violate the "read-modify-write" pattern that the MergeOperator claims to implement.

I guess I am not understanding something correctly here... can you clarify?

Reply:
> Are you saying that - If I do a single Merge(K1, V1), and there is no other preceding op for K1... then at compaction time this will be rewritten to Put(K1, V1)?

I was, but was actually mistaken. https://github.com/facebook/rocksdb/blob/fdf882ded218344c136c97daf76dfb59e4bc155f/db/merge_helper.cc#L316-L324 appears to work the way you expect. The earliest merge operand will be merged against null base value (aka `existing_value`) and the result will be written as a Put, which I believe is what you expect.

Reply:
@ajkr Okay cool! Thanks for checking that, I was beginning to doubt my sanity ;-)